### PR TITLE
Add audit data receiving implementation

### PR DESCRIPTION
**Pull request by nadundesilva**
### Proposed changes in this pull request

- Add receivers and streams for receiving audit data publisher events. (User operation events and IdP Properties Update events)
- Add siddhi extension to break a JSON map string attribute to multiple events (Claims and IdP properties are sent as a JSON map)
- Add execution plans and other artifacts for persisting latest user claims and latest IdP properties (And removing user claims upon receiving the user deletion event)

### When should this PR be merged

The [PR](https://github.com/wso2-extensions/identity-data-publisher-audit/pull/6) containing the relevant changes to the audit data publishers need to be reviewed and merged. (However, merging these changes  before that would not affect the existing features in IS Analytics)

### Follow up actions

N/A

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/org.wso2.analytics.is/src/main/java/org/wso2/analytics/is/siddhi/extension/JsonFlatMapEventDuplicatorStreamProcessor.java [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+                            }.getType());
+
+                    for (Map.Entry<String, String> entry : map.entrySet()) {
+                        StreamEvent clonedEvent = streamEventCloner.copyStreamEvent(streamEvent);
+                        complexEventPopulater
+                                .populateComplexEvent(clonedEvent, new Object[]{entry.getKey(), entry.getValue()});
+                        newStreamEventChunk.add(clonedEvent);
+                    }
+                } else {
+                    log.debug("Json Flat Map Event Duplicator stream processor " +
```
Comment (Megala21) [inline components/org.wso2.analytics.is/src/main/java/org/wso2/analytics/is/siddhi/extension/JsonFlatMapEventDuplicatorStreamProcessor.java]: I think we need to check whether debug is enabled before adding debug log

File: components/org.wso2.analytics.is/src/main/java/org/wso2/analytics/is/siddhi/extension/JsonFlatMapEventDuplicatorStreamProcessor.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    public Object[] currentState() {
+        return null;
+    }
+
+    @Override
+    public void restoreState(Object[] state) {
+        // Not required
+    }
+
+}
```
Comment (Megala21) [inline components/org.wso2.analytics.is/src/main/java/org/wso2/analytics/is/siddhi/extension/JsonFlatMapEventDuplicatorStreamProcessor.java]: Normal practice is to add an extra line at the end of the file

File: components/org.wso2.analytics.is/src/main/test/org/wso2/analytics/is/siddhi/extension/JsonFlatMapEventDuplicatorStreamProcessorTestCase.java [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+    @Before
+    public void init() {
+        inEventCount = 0;
+        removeEventCount = 0;
+        eventArrived = false;
+    }
+
+    @Test
+    public void eventDuplicatorDefaultDelimiterTest1() throws InterruptedException {
```
Comment (Megala21) [inline components/org.wso2.analytics.is/src/main/test/org/wso2/analytics/is/siddhi/extension/JsonFlatMapEventDuplicatorStreamProcessorTestCase.java]: Shall we explain on what we are actually doing with this test? Why we have 1 at the end, are we having multiple test cases?

File: components/org.wso2.analytics.is/src/main/test/org/wso2/analytics/is/siddhi/extension/JsonFlatMapEventDuplicatorStreamProcessorTestCase.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+        InputHandler inputHandler = executionPlanRuntime.getInputHandler("claimsStream");
+        executionPlanRuntime.start();
+        inputHandler.send(new Object[]{"nadund", "PRIMARY", "carbon.super",
+                "{\"urn:scim:schemas:core:1.0:userName\":\"nadund\"," +
+                        "\"http://wso2.org/claims/organization\":\"WSO2\"}"});
+        inputHandler.send(new Object[]{"john", "LEGACY", "tester.com",
+                "{\"urn:scim:schemas:core:1.0:userName\":\"john\"," +
+                        "\"http://wso2.org/claims/organization\":\"IBM\"," +
+                        "\"http://wso2.org/claims/emailaddress\":\"john@gmail.com\"}"});
+        Thread.sleep(4000);
```
Comment (Megala21) [inline components/org.wso2.analytics.is/src/main/test/org/wso2/analytics/is/siddhi/extension/JsonFlatMapEventDuplicatorStreamProcessorTestCase.java]: Normally it is not recommended to add thread.sleep in test cases, without a condition. If you are waiting for some event to occur shall we add a loop to check whether it is satisfied and add this as timeout value if possible

File: features/org.wso2.analytics.is.feature/src/main/capp/realtime/AnalyticsLatestIdPPropertiesEventSink_1.0.0/org_wso2_is_analytics_stream_LatestIdPProperties.xml [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+            <EnableScoreParam>false</EnableScoreParam>
+            <Type>STRING</Type>
+        </ColumnDefinition>
+    </TableSchema>
+    <Source>
+        <StreamId>org.wso2.is.analytics.stream.LatestIdPProperties:1.0.0</StreamId>
+    </Source>
+    <MergeSchema>false</MergeSchema>
+    <RecordStoreName>EVENT_STORE_WO</RecordStoreName>
+</EventStoreConfiguration>
```
Comment (Megala21) [inline features/org.wso2.analytics.is.feature/src/main/capp/realtime/AnalyticsLatestIdPPropertiesEventSink_1.0.0/org_wso2_is_analytics_stream_LatestIdPProperties.xml]: Add an extra line

File: features/org.wso2.analytics.is.feature/src/main/capp/realtime/IdPPropertiesUpdateAnalyticsExecutionPlan/IsAnalytics-ExecutionPlan-IdPPropertiesUpdateData.siddhiql [hunk 6]
```diff
@@ -0,0 +1,18 @@
+/* Enter a unique ExecutionPlan */
```
Comment (Megala21) [inline features/org.wso2.analytics.is.feature/src/main/capp/realtime/IdPPropertiesUpdateAnalyticsExecutionPlan/IsAnalytics-ExecutionPlan-IdPPropertiesUpdateData.siddhiql]: Missing license header

**Review by Megala21 — CHANGES_REQUESTED** [review]

File: features/org.wso2.analytics.is.feature/src/main/capp/realtime/IdPPropertiesUpdateAnalyticsExecutionPlan/IsAnalytics-ExecutionPlan-IdPPropertiesUpdateData.siddhiql (hunk 6 above)
Comment (nadundesilva) [reply features/org.wso2.analytics.is.feature/src/main/capp/realtime/IdPPropertiesUpdateAnalyticsExecutionPlan/IsAnalytics-ExecutionPlan-IdPPropertiesUpdateData.siddhiql]: The contents of this file will be shown to the user. Therefore the license header is not generally added to the .siddhiql files as far as I know.

File: components/org.wso2.analytics.is/src/main/test/org/wso2/analytics/is/siddhi/extension/JsonFlatMapEventDuplicatorStreamProcessorTestCase.java [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+                "(username string, userStoreDomain string, tenantDomain string, claims string);";
+        String query = "" +
+                "@info(name = 'query1') " +
+                "from claimsStream#isAnalytics:claimDuplicator(claims) " +
+                "select * " +
+                "insert all events into outputStream;";
+
+        ExecutionPlanRuntime executionPlanRuntime = siddhiManager.createExecutionPlanRuntime(cseEventStream + query);
+
+        executionPlanRuntime.addCallback("query1", new QueryCallback() {
```
Comment (emswbandara) [inline components/org.wso2.analytics.is/src/main/test/org/wso2/analytics/is/siddhi/extension/JsonFlatMapEventDuplicatorStreamProcessorTestCase.java]: add a new line

File: components/org.wso2.analytics.is/src/main/test/org/wso2/analytics/is/siddhi/extension/JsonFlatMapEventDuplicatorStreamProcessorTestCase.java [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+            @Override
+            public void receive(long timeStamp, Event[] inEvents, Event[] removeEvents) {
+                EventPrinter.print(timeStamp, inEvents, removeEvents);
+                if (inEvents != null) {
+                    inEventCount += inEvents.length;
+                }
+                if (removeEvents != null) {
+                    Assert.assertTrue("InEvents should arrive before RemoveEvents",
```
Comment (emswbandara) [inline components/org.wso2.analytics.is/src/main/test/org/wso2/analytics/is/siddhi/extension/JsonFlatMapEventDuplicatorStreamProcessorTestCase.java]: Better to use static imports for asserts.

File: features/org.wso2.analytics.is.feature/src/main/capp/realtime/AnalyticsLatestClaimsEventSink_1.0.0/org_wso2_is_analytics_stream_LatestClaims.xml [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+            <EnableScoreParam>false</EnableScoreParam>
+            <Type>STRING</Type>
+        </ColumnDefinition>
+    </TableSchema>
+    <Source>
+        <StreamId>org.wso2.is.analytics.stream.LatestClaims:1.0.0</StreamId>
+    </Source>
+    <MergeSchema>false</MergeSchema>
+    <RecordStoreName>EVENT_STORE_WO</RecordStoreName>
+</EventStoreConfiguration>
```
Comment (emswbandara) [inline features/org.wso2.analytics.is.feature/src/main/capp/realtime/AnalyticsLatestClaimsEventSink_1.0.0/org_wso2_is_analytics_stream_LatestClaims.xml]: add new line

File: features/org.wso2.analytics.is.feature/src/main/capp/realtime/IdPPropertiesUpdateAnalyticsExecutionPlan/artifact.xml [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+  ~ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+  ~ KIND, either express or implied.  See the License for the
+  ~ specific language governing permissions and limitations
+  ~ under the License.
+  -->
+
+<artifact name="IsAnalytics-ExecutionPlan-IdPPropertiesUpdateData" version="1.0.0" type="event/execution-plan"
+          serverRole="ComplexEventProcessor">
+    <file>IsAnalytics-ExecutionPlan-IdPPropertiesUpdateData.siddhiql</file>
+</artifact>
```
Comment (emswbandara) [inline features/org.wso2.analytics.is.feature/src/main/capp/realtime/IdPPropertiesUpdateAnalyticsExecutionPlan/artifact.xml]: add new line

File: features/org.wso2.analytics.is.feature/src/main/capp/realtime/UserDataAnalyticsExecutionPlan/artifact.xml [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+  ~ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+  ~ KIND, either express or implied.  See the License for the
+  ~ specific language governing permissions and limitations
+  ~ under the License.
+  -->
+
+<artifact name="IsAnalytics-ExecutionPlan-UserData" version="1.0.0" type="event/execution-plan"
+          serverRole="ComplexEventProcessor">
+    <file>IsAnalytics-ExecutionPlan-UserData.siddhiql</file>
+</artifact>
```
Comment (emswbandara) [inline features/org.wso2.analytics.is.feature/src/main/capp/realtime/UserDataAnalyticsExecutionPlan/artifact.xml]: add new line

File: components/org.wso2.analytics.is/src/main/java/org/wso2/analytics/is/siddhi/extension/JsonFlatMapEventDuplicatorStreamProcessor.java [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+/**
+ * Create duplicate events for a JSON string containing an array of maps.
+ */
+public class JsonFlatMapEventDuplicatorStreamProcessor extends StreamProcessor {
+    private VariableExpressionExecutor expressionExecutor;
+
+    @Override
+    protected List<Attribute> init(AbstractDefinition inputDefinition,
+                                   ExpressionExecutor[] attributeExpressionExecutors,
+                                   ExecutionPlanContext executionPlanContext) {
```
Comment (emswbandara) [inline components/org.wso2.analytics.is/src/main/java/org/wso2/analytics/is/siddhi/extension/JsonFlatMapEventDuplicatorStreamProcessor.java]: add a new line to improve code readability

File: components/org.wso2.analytics.is/src/main/test/org/wso2/analytics/is/siddhi/extension/JsonFlatMapEventDuplicatorStreamProcessorTestCase.java (hunk 4 above)
Comment (nadundesilva) [reply components/org.wso2.analytics.is/src/main/test/org/wso2/analytics/is/siddhi/extension/JsonFlatMapEventDuplicatorStreamProcessorTestCase.java]: Generally, in siddhi extension test cases, they use Thread.sleep(). We anyway have to wait for sometime to make sure that it only sends exactly the required number of events out. If we add a check to wait until the first arrived event we can't guarantee that it's not sending more than the expected number of events. So generally we should wait.

File: components/org.wso2.analytics.is/src/main/java/org/wso2/analytics/is/siddhi/extension/JsonFlatMapEventDuplicatorStreamProcessor.java [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+        this.executionPlanContext = executionPlanContext;
+
+        if (attributeExpressionExecutors.length == 1) {
+            if (attributeExpressionExecutors[0] instanceof VariableExpressionExecutor) {
+                if (attributeExpressionExecutors[0].getReturnType() == Attribute.Type.STRING) {
+                    expressionExecutor = (VariableExpressionExecutor) attributeExpressionExecutors[0];
+                } else {
+                    throw new ExecutionPlanValidationException("Json Flat Map Event Duplicator stream Processor 1st " +
```
[the remark this answers is not in the document]
Comment (nadundesilva) [reply components/org.wso2.analytics.is/src/main/java/org/wso2/analytics/is/siddhi/extension/JsonFlatMapEventDuplicatorStreamProcessor.java]: ExecutionPlanValidationException is a subclass of RuntimeException. So we do not have to add it to the method signature. What do you think?

File: components/org.wso2.analytics.is/src/main/java/org/wso2/analytics/is/siddhi/extension/JsonFlatMapEventDuplicatorStreamProcessor.java (hunk 12 above)
Comment (nadundesilva) [reply components/org.wso2.analytics.is/src/main/java/org/wso2/analytics/is/siddhi/extension/JsonFlatMapEventDuplicatorStreamProcessor.java]: Added

File: features/org.wso2.analytics.is.feature/src/main/capp/realtime/AnalyticsLatestIdPPropertiesEventSink_1.0.0/org_wso2_is_analytics_stream_LatestIdPProperties.xml (hunk 5 above)
Comment (nadundesilva) [reply features/org.wso2.analytics.is.feature/src/main/capp/realtime/AnalyticsLatestIdPPropertiesEventSink_1.0.0/org_wso2_is_analytics_stream_LatestIdPProperties.xml]: Added

File: components/org.wso2.analytics.is/src/main/java/org/wso2/analytics/is/siddhi/extension/JsonFlatMapEventDuplicatorStreamProcessor.java [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+
+    @Override
+    protected void process(ComplexEventChunk<StreamEvent> streamEventChunk, Processor nextProcessor,
+                           StreamEventCloner streamEventCloner, ComplexEventPopulater complexEventPopulater) {
+        ComplexEventChunk<StreamEvent> newStreamEventChunk = new ComplexEventChunk<>(false);
+        synchronized (this) {
+            while (streamEventChunk.hasNext()) {
+                StreamEvent streamEvent = streamEventChunk.next();
+                String jsonString = (String) expressionExecutor.execute(streamEvent);
+                if (jsonString != null) {
```
[the remark this answers is not in the document]
Comment (nadundesilva) [reply components/org.wso2.analytics.is/src/main/java/org/wso2/analytics/is/siddhi/extension/JsonFlatMapEventDuplicatorStreamProcessor.java]: Changed

File: components/org.wso2.analytics.is/src/main/test/org/wso2/analytics/is/siddhi/extension/JsonFlatMapEventDuplicatorStreamProcessorTestCase.java (hunk 7 above)
Comment (nadundesilva) [reply components/org.wso2.analytics.is/src/main/test/org/wso2/analytics/is/siddhi/extension/JsonFlatMapEventDuplicatorStreamProcessorTestCase.java]: Added

File: features/org.wso2.analytics.is.feature/src/main/capp/realtime/AnalyticsLatestClaimsEventSink_1.0.0/org_wso2_is_analytics_stream_LatestClaims.xml (hunk 9 above)
Comment (nadundesilva) [reply features/org.wso2.analytics.is.feature/src/main/capp/realtime/AnalyticsLatestClaimsEventSink_1.0.0/org_wso2_is_analytics_stream_LatestClaims.xml]: Added

File: features/org.wso2.analytics.is.feature/src/main/capp/realtime/IdPPropertiesUpdateAnalyticsExecutionPlan/artifact.xml (hunk 10 above)
Comment (nadundesilva) [reply features/org.wso2.analytics.is.feature/src/main/capp/realtime/IdPPropertiesUpdateAnalyticsExecutionPlan/artifact.xml]: Added

File: features/org.wso2.analytics.is.feature/src/main/capp/realtime/UserDataAnalyticsExecutionPlan/artifact.xml (hunk 11 above)
Comment (nadundesilva) [reply features/org.wso2.analytics.is.feature/src/main/capp/realtime/UserDataAnalyticsExecutionPlan/artifact.xml]: Added

**Review by Megala21 — APPROVED** [review]
Looks good for me.

File: components/org.wso2.analytics.is/src/main/java/org/wso2/analytics/is/siddhi/extension/JsonFlatMapEventDuplicatorStreamProcessor.java [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+                    for (Map.Entry<String, String> entry : map.entrySet()) {
+                        StreamEvent clonedEvent = streamEventCloner.copyStreamEvent(streamEvent);
+                        complexEventPopulater
+                                .populateComplexEvent(clonedEvent, new Object[]{entry.getKey(), entry.getValue()});
+                        newStreamEventChunk.add(clonedEvent);
+                    }
+                } else {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Json Flat Map Event Duplicator stream processor " +
+                                "ignored request to duplicate null object");
```
Comment (malithie) [inline components/org.wso2.analytics.is/src/main/java/org/wso2/analytics/is/siddhi/extension/JsonFlatMapEventDuplicatorStreamProcessor.java]: I really can't understand what this means :)
Ideally, what this does is cloning a new event with already read it seems.
If so, in the debug log say 'Ignored cloning a new event for empty event received' or something similar

File: components/org.wso2.analytics.is/src/main/test/org/wso2/analytics/is/siddhi/extension/EventDuplicateStreamProcessorTestCase.java [hunk 16]
```diff
@@ -43,7 +43,6 @@ public void init() {
 
     @Test
     public void eventDuplicatorDefaultDelimiterTest1() throws InterruptedException {
-
         SiddhiManager siddhiManager = new SiddhiManager();
```
Comment (malithie) [inline components/org.wso2.analytics.is/src/main/test/org/wso2/analytics/is/siddhi/extension/EventDuplicateStreamProcessorTestCase.java]: Why these lines added for readability are removed

File: components/org.wso2.analytics.is/src/main/test/org/wso2/analytics/is/siddhi/extension/JsonFlatMapEventDuplicatorStreamProcessorTestCase.java [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+
+    @Before
+    public void init() {
+        inEventCount = 0;
+        removeEventCount = 0;
+        eventArrived = false;
+    }
+
+    @Test
+    public void jsonMapToMultipleEventsTest() throws InterruptedException {
```
Comment (malithie) [inline components/org.wso2.analytics.is/src/main/test/org/wso2/analytics/is/siddhi/extension/JsonFlatMapEventDuplicatorStreamProcessorTestCase.java]: Add a new line after method signature

File: components/org.wso2.analytics.is/src/main/java/org/wso2/analytics/is/siddhi/extension/JsonFlatMapEventDuplicatorStreamProcessor.java (hunk 15 above)
Comment (nadundesilva) [reply components/org.wso2.analytics.is/src/main/java/org/wso2/analytics/is/siddhi/extension/JsonFlatMapEventDuplicatorStreamProcessor.java]: This extension doesn't duplicate events. This is similar to the event duplicator created for creating multiple events for a roles list separated by commas. The difference is that the string, in this case, is a JSON map of key-value pairs. This extension (As mentioned in the PR description) creates multiple events (one for each key-value pair). We do not pass an event into the extension to be duplicated. It's the key-value pairs that we duplicate and return as events.

Therefore, considering your suggestion as well I changed the message to "Ignored creating events from empty JSON string"

File: components/org.wso2.analytics.is/src/main/test/org/wso2/analytics/is/siddhi/extension/EventDuplicateStreamProcessorTestCase.java (hunk 16 above)
Comment (nadundesilva) [reply components/org.wso2.analytics.is/src/main/test/org/wso2/analytics/is/siddhi/extension/EventDuplicateStreamProcessorTestCase.java]: Added it back.

File: components/org.wso2.analytics.is/src/main/test/org/wso2/analytics/is/siddhi/extension/JsonFlatMapEventDuplicatorStreamProcessorTestCase.java (hunk 17 above)
Comment (nadundesilva) [reply components/org.wso2.analytics.is/src/main/test/org/wso2/analytics/is/siddhi/extension/JsonFlatMapEventDuplicatorStreamProcessorTestCase.java]: Added.

**Review by malithie — APPROVED** [review]

**Pull request merged**